### PR TITLE
Implement download slot limit

### PR DESCRIPTION
**Pull request by mathiascode**

### CODE REVIEW

**Comment by mathiascode** [comment]
It works, but more polish is needed. It's very wasteful with QueueUpload requests, mainly because check_download_queue isn't very clever at the moment.

File: pynicotine/transfers.py [hunk 1]
```diff
@@ -2085,6 +2113,28 @@ def check_queue_upload_allowed(self, user, addr, filename, real_path, msg):
...
+        if not self.allow_new_downloads():
+            return
+
+        user = None
+
+        for download in reversed(self.downloads):
+            if download.status != "Locally Queued":
+                continue
+
+            if user is None:
```
Comment (slook) [inline pynicotine/transfers.py:2129]: With this condition here, only one user will get checked. Is this intended?

File: pynicotine/transfers.py [hunk 2]
```diff
@@ -2085,6 +2113,28 @@ def check_queue_upload_allowed(self, user, addr, filename, real_path, msg):
...
+                user = download.user
+
+            elif user != download.user:
+                continue
```
Comment (slook) [inline pynicotine/transfers.py:2133]: Continuing here will result in the same user being checked again, due to the above condition.

**Comment by slook** [comment]
The the downloads list includes an item which has previously encountered the "Connection timeout" status, then the checking of Locally Queued stalls.

**Comment by slook** [comment]
> It works, but more polish is needed.

The feature does seem to serve the intended purpose as discussed in #110 quite well overall.

Given the complexity of transfers it inherently causes some quirky behaviours between peers that are inevitably going to lead to future issues, but given the refactoring that has been done hopefully this shall not be too difficult to maintain.

**Comment by slook** [comment]
Considering the pivotal changes this makes to the upload/download queuing, it would be a good idea to rebase and merge this PR as soon as possible, otherwise it will create even more work if it becomes too divergent from the master branch.

We will only find out if there are bugs in it if the testers try it out.

> more polish is needed. It's very wasteful with QueueUpload requests...

Unless you already have some specific ideas for changes that could help to address this issue, then it would be better to refine this at a later date.

**Comment by mathiascode** [comment]
It's still flawed. It can get stuck for a longer time if the uploader uses the round robin queue system.

**Comment by slook** [comment]
Might that be due to the prefer_user mechanism? Perhaps it's better just to go by the natural order for now.

**Comment by slook** [comment]
It's difficult to read the code methodically, I think it might be better to split transfers.py into downloads.py and uploads.py core modules, such as in https://github.com/slook/nicotine-plus/tree/transfers-downloads-uploads https://github.com/slook/nicotine-plus/commit/a56bd07606cfb4526664875a0257b077bd47362c , wherein it becomes much more easily apparent where optimizations are required.

**Comment by mathiascode** [comment]
@slook I agree that this should be done eventually. Keeping track of what code is related to downloads and uploads, as well as which direction messages go and which end initiated them, is difficult.

**Comment by mathiascode** [comment]
> It's difficult to read the code methodically, I think it might be better to split transfers.py into downloads.py and uploads.py core modules, such as in https://github.com/slook/nicotine-plus/tree/transfers-downloads-uploads [slook@a56bd07](https://github.com/slook/nicotine-plus/commit/a56bd07606cfb4526664875a0257b077bd47362c) , wherein it becomes much more easily apparent where optimizations are required.

Completed in https://github.com/nicotine-plus/nicotine-plus/pull/2648

**Comment by slook** [comment]
This seems like it would be totally impossible to rebase now the modules are split. Shall I close this PR?

**Comment by mathiascode** [comment]
Yes, I don't have time to work on this for now.

**Pull request closed without merging**